// mateh-ng-m-validation/esm2015/lib/ng-m-validation.component.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
let NgMValidationComponentBase = class NgMValidationComponentBase {
    constructor() { }
    ngOnInit() {
    }
};
NgMValidationComponentBase.ɵfac = function NgMValidationComponentBase_Factory(t) { return new (t || NgMValidationComponentBase)(); };
NgMValidationComponentBase.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NgMValidationComponentBase, selectors: [["mateh-ng-m-validation"]], decls: 2, vars: 0, template: function NgMValidationComponentBase_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "p");
        ɵngcc0.ɵɵtext(1, " ng-m-validation works! ");
        ɵngcc0.ɵɵelementEnd();
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgMValidationComponentBase, [{
        type: Component,
        args: [{
                selector: 'mateh-ng-m-validation',
                template: `
    <p>
      ng-m-validation works!
    </p>
  `
            }]
    }], function () { return []; }, null); })();
export { NgMValidationComponentBase };
