// mateh-ng-m-datatable/bundles/mateh-ng-m-datatable.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/forms'),require('@angular/common'),require('@angular/material/form-field'),require('@angular/material/icon'),require('@angular/material/input'),require('@angular/material/table'),require('@angular/material/sort'),require('@angular/material/paginator'),require('@angular/material/button'),require('@angular/material/menu'),exports, require('@angular/core'), require('@angular/material/paginator'), require('@angular/material/sort'), require('@angular/material/table'), require('@angular/cdk/collections'), require('rxjs/operators'), require('rxjs'), require('@angular/forms'), require('@angular/common'), require('@angular/material/form-field'), require('@angular/material/input'), require('@angular/material/icon'), require('@angular/material/button'), require('@angular/material/menu')) :
    typeof define === 'function' && define.amd ? define('mateh-ng-m-datatable', ['@angular/core','@angular/forms','@angular/common','@angular/material/form-field','@angular/material/icon','@angular/material/input','@angular/material/table','@angular/material/sort','@angular/material/paginator','@angular/material/button','@angular/material/menu','exports', '@angular/core', '@angular/material/paginator', '@angular/material/sort', '@angular/material/table', '@angular/cdk/collections', 'rxjs/operators', 'rxjs', '@angular/forms', '@angular/common', '@angular/material/form-field', '@angular/material/input', '@angular/material/icon', '@angular/material/button', '@angular/material/menu'], factory) :
    (global = global || self, factory(global.ng.core,global.ng.forms,global.ng.common,global.ng.material.formField,global.ng.material.icon,global.ng.material.input,global.ng.material.table,global.ng.material.sort,global.ng.material.paginator,global.ng.material.button,global.ng.material.menu,global['mateh-ng-m-datatable'] = {}, global.ng.core, global.ng.material.paginator, global.ng.material.sort, global.ng.material.table, global.ng.cdk.collections, global.rxjs.operators, global.rxjs, global.ng.forms, global.ng.common, global.ng.material.formField, global.ng.material.input, global.ng.material.icon, global.ng.material.button, global.ng.material.menu));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,ɵngcc6,ɵngcc7,ɵngcc8,ɵngcc9,ɵngcc10,exports, core, paginator, sort, table, collections, operators, rxjs, forms, common, formField, input, icon, button, menu) { 
function NgMDatatable_button_4_Template(rf, ctx) { if (rf & 1) {
    var _r364 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 16);
    ɵngcc0.ɵɵlistener("click", function NgMDatatable_button_4_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r364); var ctx_r363 = ɵngcc0.ɵɵnextContext(); return ctx_r363.options == null ? null : ctx_r363.options.addButton.handler(); });
    ɵngcc0.ɵɵelementStart(1, "mat-icon");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r356 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r356.options == null ? null : ctx_r356.options.addButton.icon);
} }
var _c0 = function (a0) { return { "t-right": a0 }; };
function NgMDatatable_ng_container_14_th_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "th", 20);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var c_r365 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(2, _c0, c_r365.type == "action"));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", c_r365.text, " ");
} }
function NgMDatatable_ng_container_14_td_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var row_r369 = ɵngcc0.ɵɵnextContext().$implicit;
    var c_r365 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", row_r369[c_r365.id], "");
} }
function NgMDatatable_ng_container_14_td_3_ng_container_2_button_6_Template(rf, ctx) { if (rf & 1) {
    var _r379 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 26);
    ɵngcc0.ɵɵlistener("click", function NgMDatatable_ng_container_14_td_3_ng_container_2_button_6_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r379); var a_r376 = ctx.$implicit; var row_r369 = ɵngcc0.ɵɵnextContext(2).$implicit; return a_r376.handler(row_r369); });
    ɵngcc0.ɵɵelementStart(1, "mat-icon");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "span");
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var a_r376 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("disabled", a_r376.disabled);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(a_r376.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(a_r376.text);
} }
function NgMDatatable_ng_container_14_td_3_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "button", 23);
    ɵngcc0.ɵɵelementStart(2, "mat-icon");
    ɵngcc0.ɵɵtext(3, "more_vert");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "mat-menu", null, 24);
    ɵngcc0.ɵɵtemplate(6, NgMDatatable_ng_container_14_td_3_ng_container_2_button_6_Template, 5, 3, "button", 25);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var _r374 = ɵngcc0.ɵɵreference(5);
    var c_r365 = ɵngcc0.ɵɵnextContext(2).$implicit;
    var ctx_r371 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("matMenuTriggerFor", _r374);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r371.asAction(c_r365).actions);
} }
function NgMDatatable_ng_container_14_td_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "td", 21);
    ɵngcc0.ɵɵtemplate(1, NgMDatatable_ng_container_14_td_3_ng_container_1_Template, 2, 1, "ng-container", 22);
    ɵngcc0.ɵɵtemplate(2, NgMDatatable_ng_container_14_td_3_ng_container_2_Template, 7, 2, "ng-container", 22);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var c_r365 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", c_r365.type == "text" || !c_r365.type);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", c_r365.type == "action");
} }
function NgMDatatable_ng_container_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementContainerStart(1, 17);
    ɵngcc0.ɵɵtemplate(2, NgMDatatable_ng_container_14_th_2_Template, 2, 4, "th", 18);
    ɵngcc0.ɵɵtemplate(3, NgMDatatable_ng_container_14_td_3_Template, 3, 2, "td", 19);
    ɵngcc0.ɵɵelementContainerEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var c_r365 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("matColumnDef", c_r365.id);
} }
function NgMDatatable_tr_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "tr", 27);
} }
function NgMDatatable_tr_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "tr", 28);
} }
function NgMDatatable_div_17_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 29);
    ɵngcc0.ɵɵelementStart(1, "h3", 30);
    ɵngcc0.ɵɵtext(2, " No elements ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function NgMDatatable_div_18_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 31);
    ɵngcc0.ɵɵelementStart(1, "div", 32);
    ɵngcc0.ɵɵelement(2, "div");
    ɵngcc0.ɵɵelement(3, "div", 33);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
var _c1 = function () { return [25, 50, 100, 250]; };
'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    // TODO: Replace this with your own data model type
    // export interface DataTableItem {
    //   name: string;
    //   id: number;
    // }
    /**
     * Data source for the DataTable view. This class should
     * encapsulate all logic for fetching and manipulating the displayed data
     * (including sorting, pagination, and filtering).
     */
    var DataTableDataSource = /** @class */ (function (_super) {
        __extends(DataTableDataSource, _super);
        function DataTableDataSource(data) {
            var _this = _super.call(this) || this;
            _this.data = data;
            return _this;
        }
        /**
         * Connect this data source to the table. The table will only update when
         * the returned stream emits new items.
         * @returns A stream of the items to be rendered.
         */
        DataTableDataSource.prototype.connect = function () {
            var _this = this;
            // Combine everything that affects the rendered data into one update
            // stream for the data-table to consume.
            var dataMutations = [
                rxjs.of(this.data),
                this.paginator.page,
                this.sort.sortChange,
            ];
            return rxjs.merge.apply(void 0, __spread(dataMutations)).pipe(operators.map(function () {
                return _this.getPagedData(_this.getSortedData(__spread(_this.data)));
            }));
        };
        /**
         *  Called when the table is being destroyed. Use this function, to clean up
         * any open connections or free any held resources that were set up during connect.
         */
        DataTableDataSource.prototype.disconnect = function () { };
        /**
         * Paginate the data (client-side). If you're using server-side pagination,
         * this would be replaced by requesting the appropriate data from the server.
         */
        DataTableDataSource.prototype.getPagedData = function (data) {
            var startIndex = this.paginator.pageIndex * this.paginator.pageSize;
            return data.splice(startIndex, this.paginator.pageSize);
        };
        /**
         * Sort the data (client-side). If you're using server-side sorting,
         * this would be replaced by requesting the appropriate data from the server.
         */
        DataTableDataSource.prototype.getSortedData = function (data) {
            var _this = this;
            if (!this.sort.active || this.sort.direction === "") {
                return data;
            }
            return data.sort(function (a, b) {
                var isAsc = _this.sort.direction === "asc";
                switch (_this.sort.active) {
                    case "name":
                        return compare(a["name"], b["name"], isAsc);
                    case "id":
                        return compare(+a["id"], +b["id"], isAsc);
                    default:
                        return 0;
                }
            });
        };
        return DataTableDataSource;
    }(collections.DataSource));
    /** Simple sort comparator for example ID/Name columns (for client-side sorting). */
    function compare(a, b, isAsc) {
        return (a < b ? -1 : 1) * (isAsc ? 1 : -1);
    }

    var NgMDatatable = /** @class */ (function () {
        function NgMDatatable(fb) {
            var _this = this;
            this.data = [];
            this.showSpinner = true;
            this.searchForm = fb.group({
                search: [""],
            });
            this.searchForm.valueChanges.subscribe(function (v) {
                _this.dataSource.data = _this.filter(v.search);
                _this.paginator._changePageSize(_this.paginator.pageSize);
            });
        }
        NgMDatatable.prototype.ngOnInit = function () {
            this.dataSource = new DataTableDataSource(this.data || []);
        };
        NgMDatatable.prototype.asAction = function (c) {
            return c;
        };
        NgMDatatable.prototype.filter = function (searchName) {
            return searchName && searchName.trim()
                ? this.data.filter(function (x) {
                    for (var key in x) {
                        if (x[key].toString().toLowerCase().includes(searchName.toLowerCase()))
                            return x;
                    }
                })
                : this.data;
        };
        NgMDatatable.prototype.ngAfterViewInit = function () {
            this.dataSource.data = this.data || [];
            this.dataSource.sort = this.sort;
            this.dataSource.paginator = this.paginator;
            this.table.dataSource = this.dataSource;
            this.tableBg = window.getComputedStyle(document.getElementById("klk")).background;
            this.tableColor = window.getComputedStyle(document.querySelector(".mat-header-cell")).color;
            document
                .querySelector(".table-container")
                .setAttribute("style", "background : " + this.tableBg + "; color : " + this.tableColor);
            if (this.data.length > 0)
                this.toggleLoading(false);
        };
        NgMDatatable.prototype.toggleLoading = function (v) {
            this.showSpinner = v;
            var el = document.querySelector("#loading");
            if (el)
                el.setAttribute("style", "border-color : " + this.tableColor + ";");
        };
        NgMDatatable.prototype.ngOnChanges = function (changes) {
            console.log("changes", changes);
            if (changes.data && !changes.data.firstChange) {
                this.dataSource.data = this.data || [];
                this.toggleLoading(false);
                this.paginator._changePageSize(this.paginator.pageSize);
            }
        };
        NgMDatatable.ctorParameters = function () { return [
            { type: forms.FormBuilder }
        ]; };
        __decorate([
            core.ViewChild(paginator.MatPaginator, { static: false })
        ], NgMDatatable.prototype, "paginator", void 0);
        __decorate([
            core.ViewChild(sort.MatSort, { static: false })
        ], NgMDatatable.prototype, "sort", void 0);
        __decorate([
            core.ViewChild(table.MatTable, { static: false })
        ], NgMDatatable.prototype, "table", void 0);
        __decorate([
            core.Input()
        ], NgMDatatable.prototype, "options", void 0);
        __decorate([
            core.Input()
        ], NgMDatatable.prototype, "data", void 0);
NgMDatatable.ɵfac = function NgMDatatable_Factory(t) { return new (t || NgMDatatable)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.FormBuilder)); };
NgMDatatable.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NgMDatatable, selectors: [["data-table"]], viewQuery: function NgMDatatable_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(paginator.MatPaginator, true);
        ɵngcc0.ɵɵviewQuery(sort.MatSort, true);
        ɵngcc0.ɵɵviewQuery(table.MatTable, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.paginator = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.sort = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.table = _t.first);
    } }, inputs: { data: "data", options: "options" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 21, vars: 13, consts: [[1, "table-container", "mat-elevation-z8"], [1, "table-header"], [1, "f-left"], ["mat-icon-button", "", "aria-label", "Example icon-button with a heart icon", 3, "click", 4, "ngIf"], [1, "f-right", 3, "formGroup"], [1, "example-full-width"], ["matPrefix", ""], ["formControlName", "search", "matInput", "", "placeholder", "...", "value", ""], ["mat-table", "", "matSort", "", "aria-label", "Elements", 1, "full-width-table"], [4, "ngFor", "ngForOf"], ["id", "klk", "mat-header-row", "", 4, "matHeaderRowDef"], ["mat-row", "", 4, "matRowDef", "matRowDefColumns"], ["class", "", 4, "ngIf"], ["style", "width: 100%;", "class", "d-flex justify-content-center", 4, "ngIf"], [3, "length", "pageIndex", "pageSize", "pageSizeOptions"], ["paginator", ""], ["mat-icon-button", "", "aria-label", "Example icon-button with a heart icon", 3, "click"], [3, "matColumnDef"], ["mat-header-cell", "", "mat-sort-header", "", 3, "ngClass", 4, "matHeaderCellDef"], ["mat-cell", "", 4, "matCellDef"], ["mat-header-cell", "", "mat-sort-header", "", 3, "ngClass"], ["mat-cell", ""], [4, "ngIf"], ["mat-icon-button", "", "aria-label", "Example icon-button with a menu", 1, "f-right", 3, "matMenuTriggerFor"], ["menu", "matMenu"], ["data-toggle", "modal", "data-target", "#add-serviceCats-modal", "mat-menu-item", "", 3, "disabled", "click", 4, "ngFor", "ngForOf"], ["data-toggle", "modal", "data-target", "#add-serviceCats-modal", "mat-menu-item", "", 3, "disabled", "click"], ["id", "klk", "mat-header-row", ""], ["mat-row", ""], [1, ""], [2, "text-align", "center"], [1, "d-flex", "justify-content-center", 2, "width", "100%"], [1, "lds-ripple", 2, "margin", "5px auto"], [2, "border-color", "green"]], template: function NgMDatatable_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "h2");
        ɵngcc0.ɵɵtemplate(4, NgMDatatable_button_4_Template, 3, 1, "button", 3);
        ɵngcc0.ɵɵtext(5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "form", 4);
        ɵngcc0.ɵɵelementStart(7, "mat-form-field", 5);
        ɵngcc0.ɵɵelementStart(8, "mat-icon", 6);
        ɵngcc0.ɵɵtext(9, "search");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "mat-label");
        ɵngcc0.ɵɵtext(11, "Search");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(12, "input", 7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(13, "table", 8);
        ɵngcc0.ɵɵtemplate(14, NgMDatatable_ng_container_14_Template, 4, 1, "ng-container", 9);
        ɵngcc0.ɵɵtemplate(15, NgMDatatable_tr_15_Template, 1, 0, "tr", 10);
        ɵngcc0.ɵɵtemplate(16, NgMDatatable_tr_16_Template, 1, 0, "tr", 11);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(17, NgMDatatable_div_17_Template, 3, 0, "div", 12);
        ɵngcc0.ɵɵtemplate(18, NgMDatatable_div_18_Template, 4, 0, "div", 13);
        ɵngcc0.ɵɵelement(19, "mat-paginator", 14, 15);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.options == null ? null : ctx.options.addButton);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.options == null ? null : ctx.options.title, " ");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("formGroup", ctx.searchForm);
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.options == null ? null : ctx.options.columns);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("matHeaderRowDef", ctx.options == null ? null : ctx.options.displayedColumns);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("matRowDefColumns", ctx.options == null ? null : ctx.options.displayedColumns);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.dataSource.data.length < 1 && !ctx.showSpinner);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showSpinner);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("length", ctx.dataSource == null ? null : ctx.dataSource.data.length)("pageIndex", 0)("pageSize", 50)("pageSizeOptions", ɵngcc0.ɵɵpureFunction0(12, _c1));
    } }, directives: [ɵngcc2.NgIf, ɵngcc1.ɵangular_packages_forms_forms_y, ɵngcc1.NgControlStatusGroup, ɵngcc1.FormGroupDirective, ɵngcc3.MatFormField, ɵngcc4.MatIcon, ɵngcc3.MatPrefix, ɵngcc3.MatLabel, ɵngcc1.DefaultValueAccessor, ɵngcc5.MatInput, ɵngcc1.NgControlStatus, ɵngcc1.FormControlName, ɵngcc6.MatTable, ɵngcc7.MatSort, ɵngcc2.NgForOf, ɵngcc6.MatHeaderRowDef, ɵngcc6.MatRowDef, ɵngcc8.MatPaginator, ɵngcc9.MatButton, ɵngcc6.MatColumnDef, ɵngcc6.MatHeaderCellDef, ɵngcc6.MatCellDef, ɵngcc6.MatHeaderCell, ɵngcc7.MatSortHeader, ɵngcc2.NgClass, ɵngcc6.MatCell, ɵngcc10.MatMenuTrigger, ɵngcc10._MatMenu, ɵngcc10.MatMenuItem, ɵngcc6.MatHeaderRow, ɵngcc6.MatRow], styles: ["*[_ngcontent-%COMP%]{box-sizing:border-box}.full-width-table[_ngcontent-%COMP%]{width:100%}.t-right[_ngcontent-%COMP%]{display:flex;padding:21px 0;justify-content:flex-end}.data-table[_ngcontent-%COMP%]{width:80%;margin:20px auto}.lds-ripple[_ngcontent-%COMP%]{position:relative;width:80px;height:80px}.lds-ripple[_ngcontent-%COMP%]   div[_ngcontent-%COMP%]{position:absolute;border:4px solid #4b3d3d;opacity:1;border-radius:50%;-webkit-animation:1s cubic-bezier(0,.2,.8,1) infinite lds-ripple;animation:1s cubic-bezier(0,.2,.8,1) infinite lds-ripple}.lds-ripple[_ngcontent-%COMP%]   div[_ngcontent-%COMP%]:nth-child(2){-webkit-animation-delay:-.5s;animation-delay:-.5s}@-webkit-keyframes lds-ripple{0%{top:36px;left:36px;width:0;height:0;opacity:1}100%{top:0;left:0;width:72px;height:72px;opacity:0}}@keyframes lds-ripple{0%{top:36px;left:36px;width:0;height:0;opacity:1}100%{top:0;left:0;width:72px;height:72px;opacity:0}}.table-container[_ngcontent-%COMP%]{padding:20px 5px;width:100%}.table-header[_ngcontent-%COMP%]{margin:0 40px}.f-right[_ngcontent-%COMP%]{float:right}.f-left[_ngcontent-%COMP%]{float:left}table[_ngcontent-%COMP%]{max-height:80vh}table.mat-table[_ngcontent-%COMP%]{border-spacing:0;width:100%;display:block;overflow-x:auto}.mat-table[_ngcontent-%COMP%]{width:100%;max-width:100%;display:table;border-collapse:collapse;margin:0}.mat-header-row[_ngcontent-%COMP%], .mat-row[_ngcontent-%COMP%]{display:table-row}.mat-cell[_ngcontent-%COMP%], .mat-header-cell[_ngcontent-%COMP%]{word-wrap:initial;display:table-cell;padding:0 5px;line-break:unset;width:100%;white-space:nowrap;overflow:hidden;vertical-align:middle}.mat-header-row[_ngcontent-%COMP%]{position:sticky;position:-webkit-sticky;top:0;z-index:1;background-color:inherit}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgMDatatable, [{
        type: core.Component,
        args: [{
                selector: "data-table",
                template: "<div class=\"table-container mat-elevation-z8\">\n  <div class=\"table-header\">\n    <div class=\"f-left\">\n      <h2>\n        <button\n          *ngIf=\"options?.addButton\"\n          (click)=\"options?.addButton.handler()\"\n          mat-icon-button\n          aria-label=\"Example icon-button with a heart icon\"\n        >\n          <mat-icon>{{ options?.addButton.icon }}</mat-icon>\n        </button>\n        {{ options?.title }}\n      </h2>\n    </div>\n    <form\n      class=\"mat-header-cell\"\n      [formGroup]=\"searchForm\"\n      class=\"\"\n      class=\"f-right\"\n    >\n      <mat-form-field class=\"example-full-width\">\n        <mat-icon matPrefix>search</mat-icon>\n        <mat-label>Search</mat-label>\n        <input formControlName=\"search\" matInput placeholder=\"...\" value=\"\" />\n      </mat-form-field>\n    </form>\n  </div>\n  <table mat-table class=\"full-width-table\" matSort aria-label=\"Elements\">\n    <ng-container *ngFor=\"let c of options?.columns\">\n      <ng-container [matColumnDef]=\"c.id\">\n        <th\n          [ngClass]=\"{ 't-right': c.type == 'action' }\"\n          mat-header-cell\n          *matHeaderCellDef\n          mat-sort-header\n        >\n          {{ c.text }}\n        </th>\n        <td mat-cell *matCellDef=\"let row\">\n          <ng-container *ngIf=\"c.type == 'text' || !c.type\">\n            {{ row[c.id] }}</ng-container\n          >\n          <ng-container *ngIf=\"c.type == 'action'\">\n            <button\n              class=\"f-right\"\n              mat-icon-button\n              [matMenuTriggerFor]=\"menu\"\n              aria-label=\"Example icon-button with a menu\"\n            >\n              <mat-icon>more_vert</mat-icon>\n            </button>\n            <mat-menu #menu=\"matMenu\">\n              <button\n                *ngFor=\"let a of asAction(c).actions\"\n                [disabled]=\"a.disabled\"\n                (click)=\"a.handler(row)\"\n                data-toggle=\"modal\"\n                data-target=\"#add-serviceCats-modal\"\n                mat-menu-item\n              >\n                <mat-icon>{{ a.icon }}</mat-icon>\n                <span>{{ a.text }}</span>\n              </button>\n            </mat-menu>\n          </ng-container>\n        </td>\n      </ng-container>\n    </ng-container>\n\n    <tr\n      id=\"klk\"\n      mat-header-row\n      *matHeaderRowDef=\"options?.displayedColumns\"\n    ></tr>\n    <tr mat-row *matRowDef=\"let row; columns: options?.displayedColumns\"></tr>\n  </table>\n  <div *ngIf=\"dataSource.data.length < 1 && !showSpinner\" class=\"\">\n    <h3 style=\"text-align: center;\">\n      No elements\n    </h3>\n  </div>\n  <div\n    *ngIf=\"showSpinner\"\n    style=\"width: 100%;\"\n    class=\"d-flex justify-content-center\"\n  >\n    <div style=\"margin: 5px auto;\" class=\"lds-ripple\">\n      <div></div>\n      <div style=\"border-color: green;\"></div>\n    </div>\n  </div>\n\n  <mat-paginator\n    #paginator\n    [length]=\"dataSource?.data.length\"\n    [pageIndex]=\"0\"\n    [pageSize]=\"50\"\n    [pageSizeOptions]=\"[25, 50, 100, 250]\"\n  >\n  </mat-paginator>\n</div>\n",
                styles: ["*{box-sizing:border-box}.full-width-table{width:100%}.t-right{display:flex;padding:21px 0;justify-content:flex-end}.data-table{width:80%;margin:20px auto}.lds-ripple{position:relative;width:80px;height:80px}.lds-ripple div{position:absolute;border:4px solid #4b3d3d;opacity:1;border-radius:50%;-webkit-animation:1s cubic-bezier(0,.2,.8,1) infinite lds-ripple;animation:1s cubic-bezier(0,.2,.8,1) infinite lds-ripple}.lds-ripple div:nth-child(2){-webkit-animation-delay:-.5s;animation-delay:-.5s}@-webkit-keyframes lds-ripple{0%{top:36px;left:36px;width:0;height:0;opacity:1}100%{top:0;left:0;width:72px;height:72px;opacity:0}}@keyframes lds-ripple{0%{top:36px;left:36px;width:0;height:0;opacity:1}100%{top:0;left:0;width:72px;height:72px;opacity:0}}.table-container{padding:20px 5px;width:100%}.table-header{margin:0 40px}.f-right{float:right}.f-left{float:left}table{max-height:80vh}table.mat-table{border-spacing:0;width:100%;display:block;overflow-x:auto}.mat-table{width:100%;max-width:100%;display:table;border-collapse:collapse;margin:0}.mat-header-row,.mat-row{display:table-row}.mat-cell,.mat-header-cell{word-wrap:initial;display:table-cell;padding:0 5px;line-break:unset;width:100%;white-space:nowrap;overflow:hidden;vertical-align:middle}.mat-header-row{position:sticky;position:-webkit-sticky;top:0;z-index:1;background-color:inherit}"]
            }]
    }], function () { return [{ type: ɵngcc1.FormBuilder }]; }, { data: [{
            type: core.Input
        }], paginator: [{
            type: core.ViewChild,
            args: [paginator.MatPaginator, { static: false }]
        }], sort: [{
            type: core.ViewChild,
            args: [sort.MatSort, { static: false }]
        }], table: [{
            type: core.ViewChild,
            args: [table.MatTable, { static: false }]
        }], options: [{
            type: core.Input
        }] }); })();
        return NgMDatatable;
    }());

    var NgMDatatableModule = /** @class */ (function () {
        function NgMDatatableModule() {
        }
NgMDatatableModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgMDatatableModule });
NgMDatatableModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgMDatatableModule_Factory(t) { return new (t || NgMDatatableModule)(); }, imports: [[
            common.CommonModule,
            table.MatTableModule,
            forms.ReactiveFormsModule,
            menu.MatMenuModule,
            button.MatButtonModule,
            paginator.MatPaginatorModule,
            icon.MatIconModule,
            input.MatInputModule,
            sort.MatSortModule,
            formField.MatFormFieldModule,
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgMDatatableModule, { declarations: [NgMDatatable], imports: [ɵngcc2.CommonModule, ɵngcc6.MatTableModule, ɵngcc1.ReactiveFormsModule, ɵngcc10.MatMenuModule, ɵngcc9.MatButtonModule, ɵngcc8.MatPaginatorModule, ɵngcc4.MatIconModule, ɵngcc5.MatInputModule, ɵngcc7.MatSortModule, ɵngcc3.MatFormFieldModule], exports: [NgMDatatable] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgMDatatableModule, [{
        type: core.NgModule,
        args: [{
                declarations: [NgMDatatable],
                imports: [
                    common.CommonModule,
                    table.MatTableModule,
                    forms.ReactiveFormsModule,
                    menu.MatMenuModule,
                    button.MatButtonModule,
                    paginator.MatPaginatorModule,
                    icon.MatIconModule,
                    input.MatInputModule,
                    sort.MatSortModule,
                    formField.MatFormFieldModule,
                ],
                exports: [NgMDatatable]
            }]
    }], function () { return []; }, null); })();
        return NgMDatatableModule;
    }());

    exports.DataTableDataSource = DataTableDataSource;
    exports.NgMDatatable = NgMDatatable;
    exports.NgMDatatableModule = NgMDatatableModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
