// mateh-ng-m-validation/esm5/lib/ng-m-validation.module.js
import { __decorate } from "tslib";
import { NgModule } from "@angular/core";
import { NgMValidationComponentBase } from "./ng-m-validation.component";
// import { CrossFieldErrorMatcher } from "./cross-field-error-matcher";
// import { FormComponentBase } from "./form-component-base";
import * as ɵngcc0 from '@angular/core';
var NgMValidationModule = /** @class */ (function () {
    function NgMValidationModule() {
    }
NgMValidationModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgMValidationModule });
NgMValidationModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgMValidationModule_Factory(t) { return new (t || NgMValidationModule)(); }, imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgMValidationModule, { declarations: function () { return [NgMValidationComponentBase]; }, exports: function () { return [NgMValidationComponentBase]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgMValidationModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    NgMValidationComponentBase,
                ],
                imports: [],
                exports: [
                    NgMValidationComponentBase,
                ]
            }]
    }], function () { return []; }, null); })();
    return NgMValidationModule;
}());
export { NgMValidationModule };
