// mateh-ng-m-validation/esm5/lib/ng-m-validation.component.js
import { __decorate } from "tslib";
import { Component } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var NgMValidationComponentBase = /** @class */ (function () {
    function NgMValidationComponentBase() {
    }
    NgMValidationComponentBase.prototype.ngOnInit = function () {
    };
NgMValidationComponentBase.ɵfac = function NgMValidationComponentBase_Factory(t) { return new (t || NgMValidationComponentBase)(); };
NgMValidationComponentBase.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NgMValidationComponentBase, selectors: [["mateh-ng-m-validation"]], decls: 2, vars: 0, template: function NgMValidationComponentBase_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "p");
        ɵngcc0.ɵɵtext(1, " ng-m-validation works! ");
        ɵngcc0.ɵɵelementEnd();
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgMValidationComponentBase, [{
        type: Component,
        args: [{
                selector: 'mateh-ng-m-validation',
                template: "\n    <p>\n      ng-m-validation works!\n    </p>\n  "
            }]
    }], function () { return []; }, null); })();
    return NgMValidationComponentBase;
}());
export { NgMValidationComponentBase };
