// mateh-ng-m-datatable/esm5/lib/ng-m-datatable.module.js
import { __decorate } from "tslib";
import { NgModule } from "@angular/core";
import { NgMDatatable } from "./ng-m-datatable.component";
import { MatTableModule } from "@angular/material/table";
import { MatPaginatorModule } from "@angular/material/paginator";
import { MatSortModule } from "@angular/material/sort";
import { CommonModule } from "@angular/common";
import { MatFormFieldModule } from "@angular/material/form-field";
import { MatInputModule } from "@angular/material/input";
import { MatIconModule } from "@angular/material/icon";
import { MatButtonModule } from "@angular/material/button";
import { MatMenuModule } from "@angular/material/menu";
import { ReactiveFormsModule } from "@angular/forms";
import * as ɵngcc0 from '@angular/core';
var NgMDatatableModule = /** @class */ (function () {
    function NgMDatatableModule() {
    }
NgMDatatableModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgMDatatableModule });
NgMDatatableModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgMDatatableModule_Factory(t) { return new (t || NgMDatatableModule)(); }, imports: [[
            CommonModule,
            MatTableModule,
            ReactiveFormsModule,
            MatMenuModule,
            MatButtonModule,
            MatPaginatorModule,
            MatIconModule,
            MatInputModule,
            MatSortModule,
            MatFormFieldModule,
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgMDatatableModule, { declarations: function () { return [NgMDatatable]; }, imports: function () { return [CommonModule,
        MatTableModule,
        ReactiveFormsModule,
        MatMenuModule,
        MatButtonModule,
        MatPaginatorModule,
        MatIconModule,
        MatInputModule,
        MatSortModule,
        MatFormFieldModule]; }, exports: function () { return [NgMDatatable]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgMDatatableModule, [{
        type: NgModule,
        args: [{
                declarations: [NgMDatatable],
                imports: [
                    CommonModule,
                    MatTableModule,
                    ReactiveFormsModule,
                    MatMenuModule,
                    MatButtonModule,
                    MatPaginatorModule,
                    MatIconModule,
                    MatInputModule,
                    MatSortModule,
                    MatFormFieldModule,
                ],
                exports: [NgMDatatable]
            }]
    }], function () { return []; }, null); })();
    return NgMDatatableModule;
}());
export { NgMDatatableModule };
